trino-aws-proxy-spi/src/main/java/io/trino/aws/proxy/spi/rest/S3RequestRewriter.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.aws.proxy.spi.rest;
+
+import io.trino.aws.proxy.spi.credentials.Credentials;
+
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public interface S3RequestRewriter
+{
+    S3RequestRewriter NOOP = (_, _) -> Optional.empty();
+
+    record S3RewriteResult(String finalRequestBucket, String finalRequestKey)
+    {
+        public S3RewriteResult {
+            requireNonNull(finalRequestBucket, "finalRequestBucket is null");
+            requireNonNull(finalRequestKey, "finalRequestKey is null");
+        }
+    }
+
+    Optional<S3RewriteResult> rewrite(Credentials credentials, ParsedS3Request request);
```

Comment:
Did you consider returning `Optional<ParsedS3Request>`? This would be more flexible as the plugin could adjust anything it wanted. wdyt?

Reply:
I considered this, but the `ParsedS3Request` class has lots of fields and it could become error prone. Users would need to understand the different input streams the class can have for instance.
There's also the headers logic, where we rely on some headers having been stripped out of the `ParsedS3Request` so we don't send them to S3 (for instance, `Authorization`).

I feel that it is easier to expand the `S3RewriteResult` class in the future if we want to make it more extensible, than have to consider all the possible modifications people may make to our internal parsed request record.

Wdyt?